addons/test_mail/tests/test_mail_resend.py

```diff
@@ -0,0 +1,180 @@
+        }
+        #Two users
+        cls.user1 = cls.env['res.users'].with_context(userctx).create({
+            'name': 'Employee 1',
+            'login': 'e1',
+            'email': 'e1',  # invalid email
+            'notification_type': 'email',
+            'groups_id': [(6, 0, [cls.env.ref('base.group_user').id])]
+        })
+        cls.user2 = cls.env['res.users'].with_context(userctx).create({
+            'name': 'Employee 2',
+            'login': 'e2',
+            'email': 'e2@example.com',
+            'notification_type': 'email',
+            'groups_id': [(6, 0, [cls.env.ref('base.group_user').id])]
+        })
+        #Two partner
+        cls.partner1 = cls.env['res.partner'].with_context(userctx).create({
+            'name': 'Partner 1',
+            'email': 'p1'  # invalid email
+        })
+        cls.partner2 = cls.env['res.partner'].with_context(userctx).create({
+            'name': 'Partner 2',
+            'email': 'p2@example.com'
+        })
+
+        cls.send_email_counter = int(0)
+        cls.success_email_counter = int(0)
+        @api.model
+        def send_email(self, message, *args, **kwargs):
+            cls.send_email_counter += 1
+            assert '@' in message['To'], self.NO_VALID_RECIPIENT
+            cls.success_email_counter += 1
+            return message['Message-Id']
+        cls.bus_update_failure = []
+        def sendone(self, channel, message):
+            if 'type' in message and message['type'] == 'mail_failure': 
+                    cls.bus_update_failure.append((channel, message))
+        cls.env['ir.mail_server']._patch_method('send_email', send_email)
+        cls.env['bus.bus']._patch_method('sendone', sendone)
+        cls.partners = cls.env['res.partner'].concat(cls.user1.partner_id, cls.user2.partner_id, cls.partner1, cls.partner2)
+        cls.valid_email_partners = cls.env['res.partner'].concat(cls.user2.partner_id, cls.partner2)
+        cls.invalid_email_partners = cls.partners - cls.valid_email_partners
+        cls.partner_ids = cls.partners.ids
+
+    def setUp(self):
+        super(TestMailResend, self).setUp()
+        TestMailResend.bus_update_failure = []
+        TestMailResend.send_email_counter = 0
+        TestMailResend.success_email_counter = 0
+
+    def _notif_by_partner(self, message):
+        notif = self.env['mail.notification'].search([('mail_message_id', '=', message.id)], order="res_partner_id asc")
+        self.assertEqual(self.partners, notif.mapped("res_partner_id"))
+        return notif
+
+    @classmethod
+    def tearDownClass(cls):
+        # Remove mocks
+        cls.env['ir.mail_server']._revert_method('send_email')
+        cls.env['bus.bus']._revert_method('sendone')
+        super(TestMailResend, cls).tearDownClass()
+
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mail_resend_workflow(self):
+        cls = TestMailResend
+        #missconfigured server
+        @api.model
+        def connect(**kwargs):

```

Comment:
You should try to use 
```
with patch.object(Model, 'method', ...)
```
with return_value or auto_spec + side_effect, it will be clearer I think. It is also interesting to try to use standard mock tools I think.